src/cluster/src/check.rs

```diff
@@ -109,42 +125,249 @@ pub enum CheckError {
     #[error("Minikube tunnel not found, retrying")]
     MinikubeTunnelNotFoundRetry,
 
-    // Default unhandled K8 client error
+    /// Default unhandled K8 client error
     #[error("Unhandled K8 client error")]
     UnhandledK8ClientError,
+
+    /// Kubectl not found
+    #[error("Kubectl not found")]
+    KubectlNotFoundError {
+        /// source : IoError
+        #[from]
+        source: IoError,
+    },
+
+    /// Unable to parse kubectl version
+    #[error("Unable to parse kubectl version")]
+    KubectlVersionError,
+
+    /// The installed version of Kubectl is incompatible
+    #[error("Must have kubectl version {required} or later. You have {installed}")]
+    IncompatibleKubectlVersion {
+        /// The currently-installed helm version
+        installed: String,
+        /// The minimum required helm version
+        required: String,
+    },
+
+    /// Check permissions to create k8 resources
+    #[error("Permissions to create {resource} denied")]
+    PermissionError {
+        /// Name of the resource
+        resource: String,
+    },
+
+    /// Error while fetching create permissions for a resource
+    #[error("Unable to fetch permissions")]
+    FetchPermissionError,
+
+    /// One or more pre flight checks have failed
+    #[error("Some pre-install checks have failed.")]
+    PreFlightCheckError,
 }
 
-/// Getting server hostname from K8 context
-pub fn check_cluster_server_host() -> Result<(), CheckError> {
-    let config = K8Config::load()?;
-    let context = match config {
-        K8Config::Pod(_) => return Ok(()),
-        K8Config::KubeConfig(context) => context,
-    };
+/// Captures the status of the check
+pub enum StatusCheck {
+    /// Everything seems to be working as expected, check passed
+    Working(String),
+    /// Check failed due to an error, there is no work around
+    NotWorkingNoRemediation(CheckError),
+    /// Check failed due to an error, there is a work around fix
+    NotWorking(CheckError),
+}
 
-    let cluster_context = context
-        .config
-        .current_cluster()
-        .ok_or(CheckError::NoActiveKubernetesContext)?;
-    let server_url = cluster_context.cluster.server.to_owned();
-    let url =
-        Url::parse(&server_url).map_err(|source| CheckError::BadKubernetesServerUrl { source })?;
-    let host = url
-        .host()
-        .ok_or(CheckError::MissingKubernetesServerHost)?
-        .to_string();
-    if host.is_empty() {
-        return Err(CheckError::MissingKubernetesServerHost);
+#[async_trait]
+pub trait InstallCheck: Send + Sync + 'static {
+    /// perform check, if successful return success message, if fail, return fail message
+    async fn perform_check(&self) -> StatusCheck;
```

Comment:
This should still return Result.
```
async fn perform_check(&self) -> Result<StatusCheck,Error>
```

not sure what error is. depends on check code I guess